### PR TITLE
Minor improvements to ArrayView

### PR DESCRIPTION
**Pull request by bencsikandrei**
* more constexpr
* more nodiscard
* remove unused headers

### CODE REVIEW

File: test/test_array_view.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    static constexpr int arr[expectedSize]{0, 1, 2, 3, 4};
+    static constexpr rigel::base::ArrayView<int> arrv{&arr[0], 5};
+    static_assert(arrv.size() == 5);
+    static_assert(!arrv.empty());
+    static_assert(arrv[0] == 0);
+  }
+  SECTION("c array")
+  {
+    static constexpr auto expectedSize = 5;
+    static constexpr int arr[expectedSize]{0, 1, 2, 3, 4};
```
Comment (bencsikandrei) [inline test/test_array_view.cpp:41]: I'll add some more tests tomorrow

File: src/base/array_view.hpp [hunk 2]
```diff
@@ -50,48 +48,43 @@ class ArrayView
...
     : mpData(array)
     , mSize(static_cast<size_type>(N))
   {
   }
 
   // implicit on purpose
-  ArrayView(const std::vector<T>& vec) noexcept // NOLINT
-    : mpData(vec.data())
-    , mSize(static_cast<size_type>(vec.size()))
+  template <typename Container>
```
Comment (bencsikandrei) [inline src/base/array_view.hpp:66]: this allows use with std::array, std::vector, static_vector, even string..

Comment (lethal-guitar) [reply src/base/array_view.hpp:66]: 👍🏻

File: test/test_array_view.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+TEST_CASE("Array view vector")
+{
+  const std::vector<int> vec{0, 1, 2, 3, 4};
+  const auto expectedSize = vec.size();
+  const rigel::base::ArrayView<int> arrv{vec};
+  CHECK(arrv.size() == expectedSize);
+  CHECK(!arrv.empty());
+  CHECK(*arrv.begin() == 0);
```
Comment (lethal-guitar) [inline test/test_array_view.cpp:64]: Maybe check for `std::equal(vec.begin(), vec.end(), arrv.begin())`?

File: test/test_array_view.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+RIGEL_RESTORE_WARNINGS
+
+#include <array>
+#include <vector>
+
+TEST_CASE("Array view constexpr")
+{
+  SECTION("pointer and size")
+  {
+    static constexpr auto expectedSize = 5;
```
Comment (lethal-guitar) [inline test/test_array_view.cpp:31]: Do these need to be `static`, or would `constexpr` be sufficient?

File: src/base/array_view.hpp [hunk 5]
```diff
@@ -50,48 +48,43 @@ class ArrayView
...
+  constexpr ArrayView(const Container& c) noexcept // NOLINT
+    : mpData(c.data())
+    , mSize(static_cast<size_type>(c.size()))
   {
   }
 
-  const_iterator begin() const { return mpData; }
+  constexpr const_iterator begin() const noexcept { return mpData; }
```
Comment (lethal-guitar) [inline src/base/array_view.hpp:73]: While we're at it, should we go all the way and make these also `[[nodiscard]]`? (And `front`, `back`, `data`, `operator[]`)?

**Review by lethal-guitar — COMMENTED** [review]
Good changes, and thanks for adding tests, appreciate it!

File: test/test_array_view.cpp (hunk 1 above)
Comment (bencsikandrei) [reply test/test_array_view.cpp:43]: added a bunch of tests, including ctad related ones

File: test/test_array_view.cpp (hunk 3 above)
Comment (bencsikandrei) [reply test/test_array_view.cpp]: done

File: test/test_array_view.cpp (hunk 4 above)
Comment (bencsikandrei) [reply test/test_array_view.cpp]: well, the expected size is ok being constexpr, the things requiring addresses-  like the arrays - need to also have a storage specifier. from what I understand, seems like the compiler needs to put them somewhere to actually do the computation

**Pull request closed without merging**